src/rust/scheduler/scheduler.rs

```diff
@@ -130,22 +130,33 @@ impl Scheduler {
 
     /// Insert a new task into our scheduler returning a handle corresponding to it.
     pub fn insert<F: Task>(&self, future: F) -> Option<TaskHandle> {
-        let mut task_ids: RefMut<HashMap<u64, usize>> = self.task_ids.borrow_mut();
-        // If the address space for task ids is close to half full, it will become increasingly difficult to avoid
-        // collisions, so we cap the number of tasks.
-        if task_ids.len() > MAX_NUM_TASKS {
-            panic!("Too many concurrent tasks");
-        }
+        self.panic_if_too_many_tasks();
 
-        let mut rng: RefMut<SmallRng> = self.rng.borrow_mut();
         let task_name: String = future.get_name();
-
-        // The pin slab index is the total index which can be reverse-computed in a page index and an offset within the
-        // page.
+        // The pin slab index can be reverse-computed in a page index and an offset within the page.
         let pin_slab_index: usize = self.tasks.borrow_mut().insert(Box::new(future))?;
 
-        // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
+        self.add_new_pages_to_accommodate_index_gap(pin_slab_index);
+
+        // Initialize the appropriate page offset.
+        let page_refs: Ref<Vec<WakerPageRef>> = self.page_refs.borrow();
+        let (page_ref, page_offset): (&WakerPageRef, usize) = {
+            let (page_index, page_offset) = self.get_page_index_and_offset(pin_slab_index);
+            (&page_refs[page_index], page_offset)
+        };
+        page_ref.initialize(page_offset);
+
+        let task_id = self.get_new_task_id(pin_slab_index);
+
+        trace!("insert(): name={:?}, id={:?}, pin_slab_index={:?}", task_name, task_id, pin_slab_index);
+        Some(TaskHandle::new(task_id, pin_slab_index, page_ref.clone()))
+    }
+
+    // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
+    fn get_new_task_id(&self, pin_slab_index: usize) -> u64 {
```

Comment:
Before, this happened earlier. Why change?

Reply:
There is no requisite temporal ordering within these operations, so I moved the operations to the closest place where they are used.